### ISSUE TITLE
[Bug] TensorFlow JavaCPP GPU memory leak

### ISSUE THREAD

**Issue by lanking520** [tensorflow/java#447]
During the investigation from our customer issue: https://github.com/deepjavalibrary/djl/issues/1552, we found that TensorFlow JavaCPP may have memory leak while copying from CPU to GPU devices. We managed to build a minimum reproducible code in the attachment. We appreciate for your help! :)

## How to reproduce?

All you need to do, is to find a GPU device, and run the followings after you download the attachment:

```
./gradlew run
```
and you will be able to reproduce this in less than 45 seconds.

We tested on the release of 0.4.0

## Problematic point

The copyToDevice operation will produce a new TensorHandle from the original one. However, it seemed not closed even if you do .close() operation on it. .deallocate() also doesn't help here.

```
        try (PointerScope ignored = new PointerScope()) {
            TF_Status status = TF_Status.newStatus();
            TFE_TensorHandle newHandle =
                tensorflow.TFE_TensorHandleCopyToDevice(
                    handle, eagerSessionHandle, deviceName, status);
            status.throwExceptionIfNotOK();
            return newHandle.retainReference();
        }
```

## Does CPU also have this issue?

No, we tested CPU with close for more than 3 hours and did not observe a significant memory increment.

## Output logs and test setup

You can run the project under any instances with GPU. After 13k rounds, the process will crash with OOM on GPU.

[output.log.zip](https://github.com/tensorflow/java/files/8673374/output.log.zip)
[TestTF.zip](https://github.com/tensorflow/java/files/8673375/TestTF.zip)

**Comment by Craigacp** [comment]
Does the returned `TFE_TensorHandle` have a deallocator field set? If not, you could try calling `withDeallocator` on it which should register a deallocator hook.

**Comment by saudet** [comment]
Right, `TFE_TensorHandleCopyToDevice()` is normal C function. JavaCPP has no way to know that the value returned has even been allocated somehow, so we need to call `TFE_DeleteTensorHandle()` on it. If you wish though, we can add a wrapper like `TFE_TensorHandle.copyToDevice()` that attaches a deallocator that calls `TFE_DeleteTensorHandle()` on `close()` like currently being done for `TFE_NewTensorHandle()` here: https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/c_api/AbstractTFE_TensorHandle.java#L42

**Comment by lanking520** [comment]
Maybe a warning message if no deal locator assigned to a pointer during creation will be great... I will try to run with the Delete Tensor

**Comment by saudet** [comment]
A warning doesn't make sense--this is expected behavior from C APIs. For
that to make sense, we can provide a separate API, for example, the
AbstractTFE_TensorHandle class.

**Comment by karllessard** [comment]
@saudet , I agree that simply having an endpoint called `TFE_TensorHandle.copyToDevice()` that takes care of registering the deallocator is probably the right thing to do

**Comment by saudet** [comment]
@lanking520 The convention for anything wrapped by JavaCPP is very simple:  
 1. When the name of the wrapping Java method is exactly the same as the function of the underlying C/C++ API, you'll *always* need to use it in exactly the same way as you would in C/C++. There are no exceptions. If you find a method that behaves differently than its native counterpart, that's a bug that needs to be fixed.
 2. For helper methods that are not part of original native APIs, it's *always* clearly documented that deallocators get registered, such as for the methods in the [AbstractTFE_TensorHandle](https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/c_api/AbstractTFE_TensorHandle.java) class.
